### DISCUSSION TITLE
Is there a back-transform that brings a feature-vector back to its atomic structure?

### DISCUSSION THREAD

**Discussion by JoeyTYW** [SINGROUP/dscribe#100] [General]
Hi all

Is there a way to get the atomic structure (i.e.) cartesian coordinates of the atoms) from a feature-vector?

More specifically, after instantiating a SOAP descriptor object and feeding a molecule (a .xyz file), I get a feature vector. Then using the same object (SOAP with the same parameters r_cut, n_max etc.) can I somehow feed in a feature vector and get the molecule in cartesian form back?

From what I understand about the construction of the SOAP descriptor, the partial spectrum are all composed of some kind of linear combination of spherical harmonics and gaussians which I hope is invertible and thus in principle the feature vector can be mapped back to a cartesian form?

Apologies if this is a silly or trivial question. Thanks!

**Comment by lauri-codes** [comment]
Hi @JoeyTYW.

This is a very valid question. A nice and easey inverse transform from the feature vectors to a structure does not really exist, at least for the descriptors that we have implemented. It might be possible to produce a good guess with some kind of an ML model, but because there typically is an unavoidable loss of information in the process of creating the feature vectors, there will never be a perfect mapping to the other direction.

Hope this helps.

**Reply by MCVifpen** [reply]
Hi,

Still, even if it is difficult, it would be very useful to have a way to compute $\rho^{Z} (r)$  from equation (18) in your paper in Computer Physics Communications 247 (2020) 106949.

That would allow us to "see" whether our parametrization of SOAP descriptors actually maps the electron density as we expect it. I might be wrong but, as I understand it, dscribe stores $c_{nlm}^{Z}$ coefficients in a numpy array; $g_{n} (r)$ and $Y_{lm}(\theta,\phi)$ are analytical functions. So it should be possible ...

It'd be very much appreciated if you could include this developpement in your list of TODO things, or give us some hints on how to do it.

Thanks,
MCV

**Reply by lauri-codes** [reply]
Hi @JoeyTYW,

Ok, I see. The coefficients $c_{nlm}^{Z}$ are of course computed but they exist only temporarily as they in the end multiplied and summed up over $m$ to get a rotationally invariant output. This summing is what destroys the possibility of simply recreating the density from the output, but if you store the coefficients then it's a different story.

What you are after is thus not really aligned with the primary goals of `dscribe`, i.e. creating rotationally and translational invariant output features. I could support the idea of implementing the retrieval of the coefficients as a separate function in SOAP, e.g. `get_coefficients`. But as you are currently the only user requesting this, I don't think we can allocate any time for this. You can always create a pull request where you implement this yourself. Getting those coefficients out should be fairly straightforward, the time-consuming part lies always in the testing, documentation etc.
